contents/security/passwordStorage.md

```diff
@@ -0,0 +1,120 @@
+</div>
+</popover>
+
+## What is hashing?
+
+Hashing is the act of transforming a set of data into another set of data that is difficult to reverse. As an example, suppose you wanted to multiply 2 prime numbers together, such as 17 and 11. This is easy for a computer to calculate, and it results in 187. However, suppose you were given a number that is the result of 2 prime numbers - such as 299. How do you find out which 2 prime numbers multipled together to create 299?
+
+It definitely isn't 2, because the number isn't even.  
+It isn't 3, because the digits don't sum to a multiple of 3.  
+It isn't 5, because it doesn't end in a 5 or a 0.  
+Is it 7? Well, it's hard to tell, you would have to start doing division to test.  
+ 
+There are a few shortcuts that computers and mathematicans can use to make the operation faster, but overall it is a long process that isn't efficient at all, particularly for huge numbers. Thus, **it is easy to perform the function one way, but hard to reverse**.
+
+Computers use a similar forumla to convert passwords to random strings, one that is easy to generate from the original, but would take an extremely long time to generate the original from the hash.
+
+Some examples of hashing formulas are [MD5](https://www.quora.com/How-does-the-MD5-algorithm-work) and [SHA1](https://deadhacker.com/2006/02/21/sha-1-illustrated/).
+
+### Why isn't hashing enough?
+
+Unfortunately, [rainbow tables](https://en.wikipedia.org/wiki/Rainbow_table) exist.
+
+A rainbow table is a precomputed table of hashes for some set of passwords. Basically, since the hashing function is easy to compute one-way, people build huge tables of hashes wherein the plaintext is already known, so that attempting to crack hashes becomes a simple problem of looking up the hash in the table and it's corresponding value, instead of attempting to reverse the hash. Because of this, it is very easy to crack simple hashes by simply doing a lookup.
+
+An example of a service that provides this is [Crackstation](https://crackstation.net/).
+
+Since attacks like rainbow tables exist, passwords need another layer of security.
+
+## Adding salt
+
+A salt refers to appending a string of text, unique to each user, to their password before hashing them. Since each user has a unique salt, this makes rainbow tables ineffective, as the majority of the precomputed hashes won't even contain the salt, so they wouldn't even matter anyway! In this way, the salt forces the attacker to recompute the rainbow table for each password in order to be able to effectively use it. This effectively converts the attack to <trigger for="pop:brute">brute force</trigger>, as each hash must be recomputed for each possible password. Additionally, the computed rainbow table would only be useful for that specific user, as each user would have a different salt. It could take years before a password is cracked!
```

Comment:
1. I think it's important to point out that the salt should be _random_, otherwise unknowing developers may use the username as the salt (the username is technically unique to each user after all). This would defeat the purpose of salting, since common usernames can be "rainbow-tabled".
2. We should point out that many password hashing algorithms have built-in salts (`Argon2`/`scrypt`/`bcrypt`) and a good password hashing library and algo will salt automatically.